### PR TITLE
Add `ReportedException.get_metadata_by_type`

### PR DESCRIPTION
**Pull request by mkaput**
Small change which increases code robustness in case new metadata types are added.

### CODE REVIEW

File: protostar/commands/test/test_environment_exceptions_test.py [hunk 1]
```diff
@@ -102,6 +102,24 @@ def test_matching_by_partial_error_message():
...
+        def format(self) -> str:
+            return "unused"
+
+    ex = SimpleReportedException("foo")
+
+    metadata = MockMetadata()
+    ex.metadata.append(metadata)
+
+    assert ex.get_metadata_by_type(MockMetadata) is metadata
+    assert ex.get_metadata_by_type(UnusedMockMetadata) is None
```
Comment (kasperski95) [inline protostar/commands/test/test_environment_exceptions_test.py:120]: What should happen, when somebody tries to add the same metadata twice?

**Review by kasperski95 — APPROVED** [review]

File: protostar/commands/test/test_environment_exceptions_test.py (hunk 1 above)
Comment (mkaput) [reply protostar/commands/test/test_environment_exceptions_test.py:120]: It's an undocumented programmer error. I will document that the same metadata type can only be added once.

Comment (mkaput) [reply protostar/commands/test/test_environment_exceptions_test.py:120]: Done.

File: protostar/commands/test/test_environment_exceptions.py [hunk 2]
```diff
@@ -22,9 +22,16 @@ def format(self) -> str:
...
+TMetadata = TypeVar("TMetadata", bound=ExceptionMetadata)
+
+
 class ReportedException(BaseException):
     """
     The exception used for catching unexpected errors thrown from test cases and as a base class.
+
+    Contract:
+        It is illegal to attach many same-type instances of ExceptionMetadata to the same exception
+        object.
```
Comment (kasperski95) [inline protostar/commands/test/test_environment_exceptions.py:34]: Maybe making `metadata_type` private and creating a setter with an assertion checking this restriction would be a better approach.

---

#### Hybrid Objects (Objects vs Data Structures)

> This confusion sometimes leads to unfortunate hybrid data structures that are half object and half data structure. They have functions that do significant things, and they also have either public variables or public accessors and mutators that, for all intents and purposes, make the private variables public, tempting other external functions to use those variables the way a procedural program would use a data structure.

> Such hybrids make it hard to add new functions but also make it hard to add new data structures. They are the worst of both worlds. Avoid creating them. They are indicative of a muddled design whose authors are unsure of - or worse, ignorant of - whether they need protection from functions or types.

**Review by kasperski95 — APPROVED** [review]

File: protostar/commands/test/test_environment_exceptions.py (hunk 2 above)
Comment (mkaput) [reply protostar/commands/test/test_environment_exceptions.py:34]: It's not that simple. I would have to make the `metadata` field private, expose a read only view of this list as a `metadata` property, and add a new method `attach_metadata` which performs the check. A topic for separate PR nonetheless, question: is it worth it?

**Pull request merged**